Add form-specific functionality to ModelAdmin

dwreeves opened the pull request:
I added the following attributes to `ModelAdmin`:

* `form`
* `form_base_class`
* `form_args`
* `form_columns`
* `form_excluded_columns`
* `form_overrides`

All of these attributes emulate their respective functionality in the Flask-Admin API: https://flask-admin.readthedocs.io/en/latest/api/mod_model/

This PR addresses issue #96.

--

Note that I abstracted out the `get_{x}_columns()` functionality in `ModelAdmin`. Let me know if you want me to revert this back to being repetitive.

---

~~As of writing, there is one thing currently missing from my PR: the scaffolding does _not_ make use of `column_labels`. This is something I will hope to finish up soon!~~

Let me know what other changes you want me to make.

dwreeves commented:
I added support for `form_args`, and I also pass column labels into the form.

The implementation is... not the best. I still largely retain the behavior you have implemented, which is to loop over the `str` identifiers for the columns rather than the actual Columns. I think the "Form" part of the code is starting to outgrow mere `str`s. 😬 There should be some refactoring to make this work more with Column objects rather than strings, and a few other things. Hard to explain but I think you'll see what I mean when you peek at my code.

dwreeves commented:
Just added `form_overrides` as well.

Again, this will definitely need to be refactored in the future (I already implemented a little bit of refactoring of something that was there, i.e. the `get_converter()` method).

Regardless, it should be emulating the behavior of the Flask-Admin API.

aminalaee commented:
Hey, thanks a lot for this. It generally looks very good to me.
I wish you didn't add all of them in one PR as it makes the review a bit harder. But I'll try my best to review it today 👍

aminalaee commented on sqladmin/models.py:
```diff
@@ -445,43 +532,62 @@ def get_model_attr(
+        exclude: Optional[Sequence[Union[str, InstrumentedAttribute]]] = None,
+        default: Callable[[], List[Column]] = None,
+    ) -> List[Tuple[str, Column]]:
+        """Many situations in the code call for constructing a list of columns.
+        This function generalizes this pattern for any sequence of inclusions
+        or exclusions.
```
```suggestion
        """This function generalizes constructing a list of columns
        for any sequence of inclusions or exclusions.
```

aminalaee commented on sqladmin/models.py:
````diff
@@ -283,6 +288,88 @@ class UserAdmin(ModelAdmin, model=User):
+        ```
+    """
+
+    form_base_class: ClassVar[Optional[Type[Form]]] = Form
````
Do we need the Optional here?

aminalaee commented on sqladmin/models.py:
```diff
@@ -283,6 +288,88 @@ class UserAdmin(ModelAdmin, model=User):
     """Edit view template. Default is `edit.html`."""
 
+    # Form
+    form: ClassVar[Optional[Type[Form]]] = None
```
I think technically you're right to have Optional here, but since we're not using it as arguments it won't be checked? What do you think?

dwreeves replied:
So basically, all the other Optionals I think are fine to remove. But I don't think this one should be removed.

aminalaee replied:
Yes, you're right.

aminalaee commented on sqladmin/models.py:
````diff
@@ -283,6 +288,88 @@ class UserAdmin(ModelAdmin, model=User):
+        ```
+    """
+
+    form_args: ClassVar[Optional[Dict[str, Dict[str, Any]]]] = None
````
And is Optional required here? Can we default to `{}` as this is not a function argument?

dwreeves replied:
We can avoid the `Optional` here, yes.

aminalaee commented on sqladmin/models.py:
````diff
@@ -283,6 +288,88 @@ class UserAdmin(ModelAdmin, model=User):
+        ```
+    """
+
+    form_overrides: ClassVar[Optional[Dict[str, Type[Field]]]] = None
````
And here?

aminalaee commented on sqladmin/models.py:
```diff
@@ -525,7 +631,17 @@ async def update_model(self, pk: Any, data: Dict[str, Any]) -> None:
     async def scaffold_form(self) -> Type[Form]:
-        return await get_model_form(model=self.model, engine=self.engine)
+        if self.form is not None:
+            return self.form
```
Coverage missing here.

aminalaee replied:
I don't have the right permissions, I tried to add the test:

dwreeves replied:
I don't see what test you were trying to add, but I added one called `test_form_override_scaffold` inside `test_forms.py`. Good catch on lack of coverage.

aminalaee replied:
Yes, that was exactly it.

dwreeves commented:
✅ I removed most of the `Optional` class vars. I do think that `form` should remain optional (based on the specification outlined in the Flask-Admin API), but if you have an idea of an alternate implementation, let me know.

✅ I added coverage for the override of the scaffolding.

aminalaee reviewed (approved):
Thanks a lot for the great work @dwreeves 🎉

aminalaee commented:
> Woo, all checks passed! 🎉
> 
> Thanks for all the hard work you've put into this library so far. It was lots of fun to help contribute. This is a really good implementation of an admin panel framework.
> 
> I plan on making tons of use of this library for my own projects. You might see some more PRs from me in the near future, especially as it pertains to porting more things over from Flask-Admin.

I really appreciate the effort and all sorts of contributions are welcome :)

I'm glad you'll use it and contribute to make it better.
